hamilton/lifecycle/caching.py

```diff
@@ -0,0 +1,1138 @@
+    return d
+
+
+def decode_key(context_key: str) -> tuple:
+    code_and_data_string = _decompress_string(context_key)
+    code_version, _, data_stringified = code_and_data_string.partition(",")
+    if data_stringified == "<none>":
+        dep_data_versions = {}
+    else:
+        dep_data_versions = _decode_str_dict(data_stringified)
+    return code_version, dep_data_versions
+
+
+def encode_key(code_version: str, dep_data_versions: Dict[str, str]) -> str:
+    if len(dep_data_versions.keys()) > 0:
+        dependencies_stringified = _encode_str_dict(dep_data_versions)
+    else:
+        dependencies_stringified = "<none>"
+
+    return _compress_string(f"{code_version},{dependencies_stringified}")
+
+
+class CachingBehavior(enum.Enum):
+    """Behavior applied by the caching adapter
+
+    DEFAULT:
+        Try to retrieve result from cache instead of executing the node. If the node is executed, store the result.
+        Compute the result data version and store it too.
+
+    RECOMPUTE:
+        Don't try to retrieve result from cache and always execute the node. Otherwise, behaves as default.
+        Useful when nodes are stochastic (e.g., model training) or interact with external
+        components (e.g., read from database).
+
+        .. mermaid::
+
+            sequenceDiagram
+            autonumber
+
+            participant adapter
+            participant metadata_store
+            participant result_store
+
+            note left of adapter: do_execute_node
+                adapter ->> adapter: execute node
+
+    DISABLE:
+        Node is executed as if the caching feature wasn't enabled.
+        It never tries to retrieve results. Results are never stored nor versioned.
+        Behaves like IGNORE, but the node remains a dependency for downstream nodes.
+        This means downstream cache lookup will likely fail systematically (i.e., if the cache is empty).
+
+    IGNORE:
+        Node is executed as if the caching feature wasn't enable.
+        It never tries to retrieve results. Results are never stored nor versioned.
+        IGNORE means downstream nodes will ignore this node as a dependency for lookup.
+        Ignoring clients and connections can be useful since they shouldn't directly impact the downstream results.
+    """
+
+    DEFAULT = 1
+    RECOMPUTE = 2
+    DISABLE = 3
+    IGNORE = 4
+
+    @classmethod
+    def from_string(cls, string: str) -> "CachingBehavior":
+        """Create a caching behavior from a string of the enum value. This is
+        leveraged by the ``hamilton.lifecycle.caching.SmartCacheAdapter`` and
+        the ``hamilton.function_modifiers.metadata.cache`` decorator.
+
+        .. code-block::
+
+            CachingBehavior.from_string("recompute")
+
+        """
+        try:
+            return cls[string.upper()]
+        except KeyError as e:
+            raise KeyError(f"{string} is an invalid `CachingBehavior` value") from e
+
+
+class NodeRoleInTaskExecution(enum.Enum):
```

Comment:
Add comments that this is internal and we might end up changing this (I think we can align it with the way we do tasks better, just haven't scoped that out)